commit 03e42247d2df9ac55536ce6da6c79d54a2bd508b
Author: Andrey Ulanov <andreyu@google.com>
Date:   Thu Feb 23 17:41:37 2017 -0800

    journald: When enabling audit also set audit PID.
    
    When audit PID is set kernel will not log audit messages to dmesg
    as long as that process is running.
    When the PID is set we no longer need to receive the same messages via
    multicast.

diff --git a/src/journal/journald-audit.c b/src/journal/journald-audit.c
index 83c3332ab..ad621fe6c 100644
--- a/src/journal/journald-audit.c
+++ b/src/journal/journald-audit.c
@@ -447,6 +447,14 @@ void server_process_audit_message(
                 return;
         }
 
+        /* Even though we did not join multicast group the kernel may
+         * still send us messages addressed to the group. This leads
+         * to messages being duplicated twice. So here we ignore multicast messages. */
+        if (sa->nl.nl_groups) {
+                log_debug("Audit netlink message via multicast group.");
+                return;
+        }
+
         if (!ucred || ucred->pid != 0) {
                 log_debug("Audit netlink message with invalid credentials.");
                 return;
@@ -481,12 +489,14 @@ static int enable_audit(int fd, bool b) {
                 .header.nlmsg_flags = NLM_F_REQUEST,
                 .header.nlmsg_seq = 1,
                 .header.nlmsg_pid = 0,
-                .body.mask = AUDIT_STATUS_ENABLED,
+                .body.mask = AUDIT_STATUS_ENABLED | AUDIT_STATUS_PID,
                 .body.enabled = b,
+                .body.pid = getpid(),
         };
         union sockaddr_union sa = {
                 .nl.nl_family = AF_NETLINK,
                 .nl.nl_pid = 0,
+                .nl.nl_groups = 0,
         };
         struct iovec iovec = {
                 .iov_base = &request,
@@ -518,12 +528,6 @@ int server_open_audit(Server *s) {
         int r;
 
         if (s->audit_fd < 0) {
-                static const union sockaddr_union sa = {
-                        .nl.nl_family = AF_NETLINK,
-                        .nl.nl_pid    = 0,
-                        .nl.nl_groups = AUDIT_NLGRP_READLOG,
-                };
-
                 s->audit_fd = socket(AF_NETLINK, SOCK_RAW|SOCK_CLOEXEC|SOCK_NONBLOCK, NETLINK_AUDIT);
                 if (s->audit_fd < 0) {
                         if (errno == EAFNOSUPPORT || errno == EPROTONOSUPPORT)
@@ -533,15 +537,6 @@ int server_open_audit(Server *s) {
 
                         return 0;
                 }
-
-                if (bind(s->audit_fd, &sa.sa, sizeof(sa.nl)) < 0) {
-                        log_warning_errno(errno,
-                                          "Failed to join audit multicast group. "
-                                          "The kernel is probably too old or multicast reading is not supported. "
-                                          "Ignoring: %m");
-                        s->audit_fd = safe_close(s->audit_fd);
-                        return 0;
-                }
         } else
                 fd_nonblock(s->audit_fd, 1);
 
